### ISSUE TITLE
Limit data read from socket or pass back data the socket needs (StartTLS)

### ISSUE THREAD

**Issue by benjamg** [envoyproxy/envoy#18878]
I'm not overly sure this is a new feature or a bug in the way the system currently works with the StartTLS socket.

This issue occurred while handling MySQL data; the protocol dictates that an instruction is set to switch on encryption and then the TLS handshake is started, however if the downstream client has combined the two TCP packets for whatever reason then Envoy will pass both the instruction and the start of the handshake to the filter manager and so into the filter chain. This means triggering a startSecureTransport() call on the connection will just idle while it waits forever for a handshake packet.

If there was a way to pass the handshake back to the socket layer, or a way to limit the data read from the socket until this handshake phase has past it would avoid this issue.

**Comment by benjamg** [comment]
A brief conversion that lead to me raising this issue can be found on the dev group at https://groups.google.com/g/envoy-dev/c/hU2vxixuuuE. Please let me know If I need to clarify what I am asking for.

**Comment by ggreenway** [comment]
As I mentioned in the email, one possibility is something similar to what was discussed in #17395. In that case it was so that listener filters could do a read instead of a PEEK to avoid re-reading the data. If we had something like that (a way to put data back into the IoHandle to be used for the next read), `startSecureTransport()` could take a buffer of data that is destined for the TLS transport socket.

@mattklein123 any other ideas on how to fix this?

**Comment by mattklein123** [comment]
> @mattklein123 any other ideas on how to fix this?

Nothing comes to mind other than your idea which makes sense to me. I think having `startSecureTransport()` take extra bytes to push back on, and then having a way to push them back onto the receive queue for the socket makes sense to me. cc @soulxu.

**Comment by soulxu** [comment]
> > @mattklein123 any other ideas on how to fix this?
> 
> Nothing comes to mind other than your idea which makes sense to me. I think having `startSecureTransport()` take extra bytes to push back on, and then having a way to push them back onto the receive queue for the socket makes sense to me. cc @soulxu.

Yes, that works since I write PoC for listener filter. And we have to push the data into IoHandle, since there is no way to inject into BoringSSL directly, have to through the BIO, then through the ioHandle.

I'm also thinking the nested network filter may have the same requirement, how the nested filter to do the handshake with data already read out? cc @kyessenov

I'm not similar with MySQL protocol enough, if the starttls happened early, is it could be done by the listener filter? But listener filter is very limit, since it doesn't support response data.

**Comment by cpakulski** [comment]
I am aware of this problem and it requires some extra API between filter and StartTLS transport socket. Specifically, MySQL filter has to tell StartTLS socket: "read only X bytes and then stop (buffer) remaining data." That X bytes contains info whether SSL should be turned on or not. If SSL should be turned on, filter calls startSecureTransport (to switch to TLS) and instructs StartTLS transport socket to forward buffered data. That buffered data contains TLS handshake.

**Comment by mattklein123** [comment]
> Specifically, MySQL filter has to tell StartTLS socket: "read only X bytes and then stop (buffer) remaining data."

Yeah this is another way of doing it, but I suspect this will be more difficult in practice than allowing the data to get pushed back on manually. I agree that either would work in theory.
